File: runtime/src/storage/mkvs/urkel/tree/sync.rs
use std::{any::Any, sync::Arc};

use failure::{Error, Fallible};
use io_context::Context;

use crate::storage::mkvs::urkel::{cache::*, sync::*, tree::*};

impl ReadSync for UrkelTree {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn get_subtree(
        &mut self,
        ctx: Context,
        root: Root,
        id: NodeID,
        max_depth: Depth,
    ) -> Fallible<Subtree> {
        let ctx = ctx.freeze();
        let pending_root = self.cache.borrow().get_pending_root();
        if root != self.cache.borrow().get_sync_root() {
            return Err(SyncerError::InvalidRoot.into());
        }
        if !pending_root.borrow().clean {
            return Err(SyncerError::DirtyRoot.into());
        }

        let (subtree_root, bd) = self.cache.borrow_mut().deref_node_id(&ctx, id)?;
        if subtree_root.borrow().is_null() {
            return Err(SyncerError::NodeNotFound.into());
        }

        // path corresponds to already navigated prefix of the key up to bd bits.
        let (path, _) = id.path.split(bd, id.path.bit_length());
        let mut subtree = Subtree::new();

        let root_ptr = self._get_subtree(
            &ctx,
            subtree_root,
            bd,
            &path,
            &mut subtree,
            0,
            max_depth,
            if id.path.len() > 0 {
                id.path.get_bit(bd)
            } else {
                false
            },
        )?;
        subtree.root = root_ptr;
        if !subtree.root.valid {
            Err(SyncerError::InvalidRoot.into())
        } else {
            Ok(subtree)
        }
    }

    fn get_path(
        &mut self,
        ctx: Context,
        root: Root,
        key: &Key,
        start_bit_depth: Depth,
    ) -> Fallible<Subtree> {
        let ctx = ctx.freeze();
        if root != self.cache.borrow().get_sync_root() {
            return Err(SyncerError::InvalidRoot.into());
        }
        if !self.cache.borrow().get_pending_root().borrow().clean {
            return Err(SyncerError::DirtyRoot.into());
        }

        let (subtree_root, bd) = self
            .cache
            .borrow_mut()
            .deref_node_id(
                &ctx,
                NodeID {
                    path: key,
                    bit_depth: start_bit_depth,
                },
            )
            .map_err(|_| Error::from(SyncerError::NodeNotFound))?;

        let mut subtree = Subtree::new();

        // path corresponds to already navigated prefix of the key up to bd bits.
        let (path, _) = key.split(bd, key.bit_length());
        subtree.root = self._get_path(
            &ctx,
            subtree_root,
            bd,
            &path,
            Some(key),
            &mut subtree,
            if key.len() > 0 {
                key.get_bit(bd)
            } else {
                false
            },
        )?;
        if !subtree.root.valid {
            Err(SyncerError::InvalidRoot.into())
        } else {
            Ok(subtree)
        }
    }

    fn get_node(&mut self, ctx: Context, root: Root, id: NodeID) -> Fallible<NodeRef> {
        let ctx = ctx.freeze();
        if root != self.cache.borrow().get_sync_root() {
            Err(SyncerError::InvalidRoot.into())
        } else if !self.cache.borrow().get_pending_root().borrow().clean {
            Err(SyncerError::DirtyRoot.into())
        } else {
            let ptr = self
                .cache
                .borrow_mut()
                .deref_node_id(&ctx, id)
                .map_err(|_| Error::from(SyncerError::NodeNotFound))?
                .0;
            let node = self
                .cache
                .borrow_mut()
                .deref_node_ptr(&ctx, id, ptr, None)
                .map_err(|_| Error::from(SyncerError::NodeNotFound))?;
            Ok(node.unwrap().borrow().extract())
        }
    }
}

impl UrkelTree {
    fn _get_subtree(
        &mut self,
        ctx: &Arc<Context>,
        ptr: NodePtrRef,
        bit_depth: Depth,
        path: &Key,
        st: &mut Subtree,
        depth: Depth,
        max_depth: Depth,
        right: bool,
    ) -> Fallible<SubtreePointer> {
        let node_ref = self.cache.borrow_mut().deref_node_ptr(
            ctx,
            NodeID {
                path: &path.append_bit(bit_depth, right),
                bit_depth: bit_depth + 1,
            },
            ptr.clone(),
            None,
        )?;
        let node_ref = match node_ref {
            None => {
                return Ok(SubtreePointer {
                    index: SubtreeIndex::invalid(),
                    valid: true,
                    ..Default::default()
                })
            }
            Some(node_ref) => node_ref,
        };

        if depth >= max_depth {
            // Nodes at max_depth are always full nodes.
            let idx = st.add_full_node(node_ref.borrow().extract())?;
            return Ok(SubtreePointer {
                index: idx,
                full: true,
                valid: true,
            });
        }

        match classify_noderef!(node_ref) {
            NodeKind::None => unreachable!(),
            NodeKind::Internal => {
                let mut summary = InternalNodeSummary {
                    ..Default::default()
                };

                summary.label = noderef_as!(node_ref, Internal).label.clone();
                summary.label_bit_length = noderef_as!(node_ref, Internal).label_bit_length;

                let new_path = path.merge(
                    bit_depth,
                    &noderef_as!(node_ref, Internal).label,
                    noderef_as!(node_ref, Internal).label_bit_length,
                );
                summary.leaf_node = self._get_subtree(
                    ctx,
                    noderef_as!(node_ref, Internal).leaf_node.clone(),
                    bit_depth + noderef_as!(node_ref, Internal).label_bit_length,
                    &new_path,
                    st,
                    depth,
                    max_depth,
                    false,
                )?;
                summary.left = self._get_subtree(
                    ctx,
                    noderef_as!(node_ref, Internal).left.clone(),
                    bit_depth + noderef_as!(node_ref, Internal).label_bit_length,
                    &new_path,
                    st,
                    depth + 1,
                    max_depth,
                    false,
                )?;
                summary.right = self._get_subtree(
                    ctx,
                    noderef_as!(node_ref, Internal).right.clone(),
                    bit_depth + noderef_as!(node_ref, Internal).label_bit_length,
                    &new_path,
                    st,
                    depth + 1,
                    max_depth,
                    true,
                )?;

                let idx = st.add_summary(&summary)?;
                return Ok(SubtreePointer {
                    index: idx,
                    valid: true,
                    ..Default::default()
                });
            }
            NodeKind::Leaf => {
                let idx = st.add_full_node(node_ref.borrow().extract())?;
                return Ok(SubtreePointer {
                    index: idx,
                    full: true,
                    valid: true,
                });
            }
        };
    }

    fn _get_path(
        &mut self,
        ctx: &Arc<Context>,
        ptr: NodePtrRef,
        bit_depth: Depth,
        path: &Key,
        key: Option<&Key>,
        st: &mut Subtree,
        right: bool,
    ) -> Fallible<SubtreePointer> {
        let ext_path = path.append_bit(bit_depth, right);
        let node_ref = self.cache.borrow_mut().deref_node_ptr(
            ctx,
            NodeID {
                path: &ext_path,
                bit_depth: bit_depth + 1,
            },
            ptr.clone(),
            Some(&ext_path),
        )?;
        let node_ref = match node_ref {
            None => {
                return Ok(SubtreePointer {
                    index: SubtreeIndex::invalid(),
                    valid: true,
                    ..Default::default()
                })
            }
            Some(node_ref) => node_ref,
        };

        if key.is_none() {
            // Off-path nodes are always full nodes.
            let idx = st.add_full_node(node_ref.borrow().extract())?;
            return Ok(SubtreePointer {
                index: idx,
                full: true,
                valid: true,
            });
        }
        let key = key.expect("key is not none");

        match classify_noderef!(node_ref) {
            NodeKind::None => unreachable!(),
            NodeKind::Internal => {
                // Determine which subtree is off-path.
                let (mut left_key, mut right_key) = (None, None);
                if bit_depth + noderef_as!(node_ref, Internal).label_bit_length < key.bit_length() {
                    if key.get_bit(bit_depth + noderef_as!(node_ref, Internal).label_bit_length) {
                        // Left subtree is off-path.
                        right_key = Some(key)
                    } else {
                        // Right subtree is off-path.
                        left_key = Some(key)
                    }
                }

                let mut summary = InternalNodeSummary {
                    ..Default::default()
                };

                summary.label = noderef_as!(node_ref, Internal).label.clone();
                summary.label_bit_length = noderef_as!(node_ref, Internal).label_bit_length;

                let new_path = path.merge(
                    bit_depth,
                    &noderef_as!(node_ref, Internal).label,
                    noderef_as!(node_ref, Internal).label_bit_length,
                );
                summary.leaf_node = self._get_path(
                    ctx,
                    noderef_as!(node_ref, Internal).leaf_node.clone(),
                    bit_depth + noderef_as!(node_ref, Internal).label_bit_length,
                    &new_path,
                    Some(&path),
                    st,
                    false,
                )?;
                summary.left = self._get_path(
                    ctx,
                    noderef_as!(node_ref, Internal).left.clone(),
                    bit_depth + noderef_as!(node_ref, Internal).label_bit_length,
                    &new_path,
                    left_key,
                    st,
                    false,
                )?;
                summary.right = self._get_path(
                    ctx,
                    noderef_as!(node_ref, Internal).right.clone(),
                    bit_depth + noderef_as!(node_ref, Internal).label_bit_length,
                    &new_path,
                    right_key,
                    st,
                    true,
                )?;

                let idx = st.add_summary(&summary)?;
                return Ok(SubtreePointer {
                    index: idx,
                    full: false,
                    valid: true,
                    ..Default::default()
                });
            }
            NodeKind::Leaf => {
                // All encountered leaves are always full nodes.
                let idx = st.add_full_node(node_ref.borrow().extract())?;
                return Ok(SubtreePointer {
                    index: idx,
                    full: true,
                    valid: true,
                });
            }
        };
    }
}

File: go/storage/mkvs/urkel/sync.go
package urkel

import (
	"context"

	"github.com/pkg/errors"

	"github.com/oasislabs/ekiden/go/storage/mkvs/urkel/node"
	"github.com/oasislabs/ekiden/go/storage/mkvs/urkel/syncer"
)

var _ syncer.ReadSyncer = (*Tree)(nil)

// GetSubtree retrieves a compressed subtree summary of the given node
// under the given root up to the specified depth.
//
// It is the responsibility of the caller to validate that the subtree
// is correct and consistent.
func (t *Tree) GetSubtree(ctx context.Context, root node.Root, id node.ID, maxDepth node.Depth) (*syncer.Subtree, error) {
	t.cache.Lock()
	defer t.cache.Unlock()

	if t.cache.isClosed() {
		return nil, ErrClosed
	}
	if !root.Equal(&t.cache.syncRoot) {
		return nil, syncer.ErrInvalidRoot
	}
	if !t.cache.pendingRoot.IsClean() {
		return nil, syncer.ErrDirtyRoot
	}

	// Extract the node that is at the root of the subtree.
	subtreeRoot, bd, err := t.cache.derefNodeID(ctx, id)
	if err != nil {
		return nil, syncer.ErrNodeNotFound
	}

	// path corresponds to already navigated prefix of the key up to bd bits.
	path, _ := id.Path.Split(bd, id.Path.BitLength())

	st := &syncer.Subtree{}
	right := false
	if len(id.Path) > 0 {
		right = id.Path.GetBit(bd)
	}
	rootPtr, err := t.doGetSubtree(ctx, subtreeRoot, bd, path, st, 0, maxDepth, right)
	if err != nil {
		return nil, errors.Wrap(err, "urkel: failed to get subtree")
	}
	st.Root = rootPtr
	if !st.Root.Valid {
		return nil, syncer.ErrInvalidRoot
	}

	return st, nil
}

func (t *Tree) doGetSubtree(
	ctx context.Context,
	ptr *node.Pointer,
	bitDepth node.Depth,
	path node.Key,
	st *syncer.Subtree,
	depth node.Depth,
	maxDepth node.Depth,
	right bool,
) (syncer.SubtreePointer, error) {
	// Abort in case the context is cancelled.
	select {
	case <-ctx.Done():
		return syncer.SubtreePointer{}, ctx.Err()
	default:
	}

	nd, err := t.cache.derefNodePtr(ctx, node.ID{Path: path.AppendBit(bitDepth, right), BitDepth: bitDepth}, ptr, nil)
	if err != nil {
		return syncer.SubtreePointer{}, err
	}
	if nd == nil {
		return syncer.SubtreePointer{Index: syncer.InvalidSubtreeIndex, Valid: true}, nil
	}

	if depth >= maxDepth {
		// Nodes at maxDepth are always full nodes.
		idx, err := st.AddFullNode(nd.Extract())
		if err != nil {
			return syncer.SubtreePointer{}, err
		}
		return syncer.SubtreePointer{Index: idx, Full: true, Valid: true}, nil
	}

	switch n := nd.(type) {
	case *node.InternalNode:
		// Record internal node summary.
		s := syncer.InternalNodeSummary{}

		// To traverse subtrees resize path bit vector, if needed.
		path = path.Merge(bitDepth, n.Label, n.LabelBitLength)

		s.Label = n.Label
		s.LabelBitLength = n.LabelBitLength

		bitLength := bitDepth + n.LabelBitLength
		newPath := path.Merge(bitDepth, n.Label, n.LabelBitLength)

		// Leaf node.
		leafNodePtr, err := t.doGetSubtree(ctx, n.LeafNode, bitLength, newPath, st, depth, maxDepth, false)
		if err != nil {
			return syncer.SubtreePointer{}, err
		}
		s.LeafNode = leafNodePtr

		// Left subtree.
		leftPtr, err := t.doGetSubtree(ctx, n.Left, bitLength, newPath, st, depth+1, maxDepth, false)
		if err != nil {
			return syncer.SubtreePointer{}, err
		}
		s.Left = leftPtr

		// Right subtree.
		rightPtr, err := t.doGetSubtree(ctx, n.Right, bitLength, newPath, st, depth+1, maxDepth, true)
		if err != nil {
			return syncer.SubtreePointer{}, err
		}
		s.Right = rightPtr

		idx, err := st.AddSummary(s)
		if err != nil {
			return syncer.SubtreePointer{}, err
		}

		return syncer.SubtreePointer{Index: idx, Valid: true}, nil
	case *node.LeafNode:
		// All encountered leaves are always full nodes.
		idx, err := st.AddFullNode(nd.Extract())
		if err != nil {
			return syncer.SubtreePointer{}, err
		}

		return syncer.SubtreePointer{Index: idx, Full: true, Valid: true}, nil
	default:
		panic("urkel: invalid node type")
	}
}

// GetPath retrieves a compressed path summary for the given key under
// the given root, starting at the given bit depth.
//
// It is the responsibility of the caller to validate that the subtree
// is correct and consistent.
func (t *Tree) GetPath(ctx context.Context, root node.Root, key node.Key, startBitDepth node.Depth) (*syncer.Subtree, error) {
	t.cache.Lock()
	defer t.cache.Unlock()

	if t.cache.isClosed() {
		return nil, ErrClosed
	}
	if !root.Equal(&t.cache.syncRoot) {
		return nil, syncer.ErrInvalidRoot
	}
	if !t.cache.pendingRoot.IsClean() {
		return nil, syncer.ErrDirtyRoot
	}

	subtreeRoot, bd, err := t.cache.derefNodeID(ctx, node.ID{Path: key, BitDepth: startBitDepth})
	if err != nil {
		return nil, syncer.ErrNodeNotFound
	}

	st := &syncer.Subtree{}

	// TODO: Make this a GetPath argument.
	opts := syncer.PathOptions{
		IncludeSiblings: true,
	}
	rootPtr, err := t.doGetPath(ctx, subtreeRoot, bd, key, false, opts, st)
	if err != nil {
		return nil, errors.Wrap(err, "urkel: failed to get path")
	}
	st.Root = rootPtr
	if !st.Root.Valid {
		return nil, syncer.ErrInvalidRoot
	}

	return st, nil
}

func (t *Tree) doGetPath(
	ctx context.Context,
	ptr *node.Pointer,
	bitDepth node.Depth,
	key node.Key,
	stop bool,
	opts syncer.PathOptions,
	st *syncer.Subtree,
) (syncer.SubtreePointer, error) {
	// Abort in case the context is cancelled.
	select {
	case <-ctx.Done():
		return syncer.SubtreePointer{}, ctx.Err()
	default:
	}

	nd, err := t.cache.derefNodePtr(ctx, node.ID{Path: key, BitDepth: bitDepth}, ptr, key)
	if err != nil {
		return syncer.SubtreePointer{}, err
	}
	if nd == nil {
		return syncer.SubtreePointer{Index: syncer.InvalidSubtreeIndex, Valid: true}, nil
	}

	// All nodes on the path are always full nodes.
	idx, err := st.AddFullNode(nd.Extract())
	if err != nil {
		return syncer.SubtreePointer{}, err
	}

	if !stop {
		switch n := nd.(type) {
		case *node.InternalNode:
			bitLength := bitDepth + n.LabelBitLength

			// If lookup key ends at this node, we don't need to do anything more as
			// the full internal node already includes the leaf.
			if key.BitLength() <= bitLength {
				if opts.IncludeSiblings {
					_, err = t.doGetPath(ctx, n.Left, bitLength, key.AppendBit(bitLength, false), true, opts, st)
					if err != nil {
						return syncer.SubtreePointer{}, err
					}

					_, err = t.doGetPath(ctx, n.Right, bitLength, key.AppendBit(bitLength, true), true, opts, st)
					if err != nil {
						return syncer.SubtreePointer{}, err
					}
				}
				break
			}

			// Continue recursively based on a bit value.
			if key.GetBit(bitLength) {
				_, err = t.doGetPath(ctx, n.Right, bitLength, key, false, opts, st)
				if err != nil {
					return syncer.SubtreePointer{}, err
				}

				if opts.IncludeSiblings {
					ckey, _ := key.Split(bitLength, key.BitLength())
					_, err = t.doGetPath(ctx, n.Left, bitLength, ckey.AppendBit(bitLength, false), true, opts, st)
				}
			} else {
				_, err = t.doGetPath(ctx, n.Left, bitLength, key, false, opts, st)
				if err != nil {
					return syncer.SubtreePointer{}, err
				}

				if opts.IncludeSiblings {
					ckey, _ := key.Split(bitLength, key.BitLength())
					_, err = t.doGetPath(ctx, n.Right, bitLength, ckey.AppendBit(bitLength, true), true, opts, st)
				}
			}
			if err != nil {
				return syncer.SubtreePointer{}, err
			}
		}
	}

	return syncer.SubtreePointer{Index: idx, Full: true, Valid: true}, nil
}

// GetNode retrieves a specific node under the given root.
//
// It is the responsibility of the caller to validate that the node
// is consistent. The node's cached hash should be considered invalid
// and must be recomputed locally.
func (t *Tree) GetNode(ctx context.Context, root node.Root, id node.ID) (node.Node, error) {
	t.cache.Lock()
	defer t.cache.Unlock()

	if t.cache.isClosed() {
		return nil, ErrClosed
	}
	if !root.Equal(&t.cache.syncRoot) {
		return nil, syncer.ErrInvalidRoot
	}
	if !t.cache.pendingRoot.IsClean() {
		return nil, syncer.ErrDirtyRoot
	}
	ptr, _, err := t.cache.derefNodeID(ctx, id)
	if err != nil {
		return nil, syncer.ErrNodeNotFound
	}
	nd, err := t.cache.derefNodePtr(ctx, id, ptr, nil)
	if err != nil {
		return nil, syncer.ErrNodeNotFound
	}

	return nd.Extract(), nil
}
